dask_expr/_shuffle.py

```diff
@@ -618,7 +618,25 @@ def operation(cls, df, index, name: str, npartitions: int):
         return df.assign(**{name: index})
 
 
-class SetIndex(Expr):
+class BaseSetIndexSortValues(Expr):
+    def _divisions(self):
+        if self.user_divisions is not None:
+            return self.user_divisions
+        divisions, mins, maxes, presorted = _calculate_divisions(
+            self.frame, self.other, self.npartitions, self.ascending
+        )
+        if presorted:
+            divisions = mins.copy() + [maxes[-1]]
+        return divisions
+
+    @property
+    def npartitions(self):
+        if self.operand("npartitions") is not None:
+            return self.operand("npartitions")
+        return self.frame.npartitions
```

Comment:
```suggestion
        return self.operand("npartitions") or self.frame.npartitions
```
Minor preference, but happy either way.